File: aoc-2015/day-20/day_20.cpp
#include <cmath>
#include <fstream>
#include <functional>
#include <iostream>

int part1_present_sum(const int house)
{
    const int d = std::sqrt(house) + 1;
    int s = 0;
    for (int i = 1; i <= d; i++)
        if (house % i == 0)
            s += i + house / i;
    return 10 * s;
}

int part2_present_sum(const int house)
{
    const int d = std::sqrt(house) + 1;
    int s = 0;
    for (int i = 1; i <= d; i++)
        if (house % i == 0)
            s += (i <= 50 ? house / i : 0) + (house / i <= 50 ? i : 0);
    return 11 * s;
}

int find_first_house(const int target, const std::function<int(int)>& present_function)
{
    for (int n = 1;; n++)
        if (present_function(n) >= target)
            return n;
}

int main()
{
    std::ifstream input_file {"input.txt"};
    int target;
    input_file >> target;

    std::cout << "Part 1: " << find_first_house(target, part1_present_sum) << std::endl;
    std::cout << "Part 2: " << find_first_house(target, part2_present_sum) << std::endl;
}

File: aoc-2015/day-20/day_20.hs
-- stack --resolver lts-16.27 script --optimize

-- This solution is largely stolen from a solution posted in the Megathread.
-- Solved in C++ mainly, this is much slower but really nice, so I kept it for
-- reference. See original:
-- https://www.reddit.com/r/adventofcode/comments/3xjpp2/day_20_solutions/cy5ou98

import           Data.List                      ( findIndex
                                                , nub
                                                )
import           Data.Maybe                     ( fromJust )
import           Text.Printf                    ( printf )

main :: IO ()
main = do
  content <- readFile "input.txt"
  let target = read content :: Int
  printf "Part 1: %d\n" $ findFirstHouse target presents
  printf "Part 2: %d\n" $ findFirstHouse target presents2

isqrt :: Int -> Int
isqrt = floor . sqrt . fromIntegral

factors :: Int -> [Int]
factors n =
  nub
    . concat
    $ [ [p, q] | p <- [1 .. isqrt (n)], let (q, r) = divMod n p, r == 0 ]

presents :: Int -> Int
presents = (10 *) . sum . factors

presents2 :: Int -> Int
presents2 n = (11 *) . sum . filter (> (n - 1) `div` 50) . factors $ n

findFirstHouse :: Int -> (Int -> Int) -> Int
findFirstHouse target presentCounter =
  fromJust . findIndex (>= target) . map presentCounter $ [0 ..]

